### PR TITLE
DOCKER-793: First pass for NFS shared volumes

### PR DESCRIPTION
**Pull request by misterdjules**
This is a first pass that allows joyent/sdc-docker to provide [experimental support for NFS shared  modules](https://github.com/joyent/sdc-docker/pull/78). My plan is to add and update tests when the volumes API stabilizes a bit.

### CODE REVIEW

File: lib/volapi.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+util.inherits(VOLAPI, RestifyClient);
+
+
+VOLAPI.prototype.close = function close() {
+    this.client.close();
+};
+
+VOLAPI.prototype.createVolume =
+    function createVolume(params, options, callback) {
```
Comment (trentm) [inline lib/volapi.js]: Hrm. I'd like to propose that some of us talk (tomorrow?) about picking a preference for function signature for clients, at least for new ones. My preference has moved to wanting `function theName(opts, cb)` for all methods. Here I mean that there are two arguments: the first is options, the second is the callback. I notice that we aren't that consistent (surprise surprise!).

(A separate less important discussion is the what to call those vars: `opts` vs `options`, `callback` vs `cb`.)

File: lib/volapi.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        if (typeof (options) === 'function') {
+            callback = options;
```
Comment (trentm) [inline lib/volapi.js]: I'm not a fan of this double-indentation style just because the function name and/or args was long. That width for just the start of the method should impact indentation for all the lines of the function,IMHO.

File: lib/volapi.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
+var util = require('util');
```
Comment (trentm) [inline lib/volapi.js]: license and copyright block missing?

Please have two blocks of requires: external packages and then local libs.

File: lib/volapi.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+var assert = require('assert-plus');
+
+var RestifyClient = require('./restifyclient');
+
+function VOLAPI(options) {
+    RestifyClient.call(this, options);
+    this.url = options.url;
+}
```
Comment (trentm) [inline lib/volapi.js]: I'd love to move to requiring user-agent in our clients. That could be a separate PR, tho.

File: lib/volapi.js (hunk 1 above)
Comment (trentm) [inline lib/volapi.js]: Also discussable is whether all client methods should be strict in validating input options. I'm leaning towards *yes*.

File: lib/volapi.js (hunk 4 above)
Comment (trentm) [inline lib/volapi.js]: Might consider requiring specifying an API version -- even if volapi doesn't do API versioning yet. Require users to specify "acceptVersion: ~1" or explicitly "acceptVersion: *" and it makes moving to api versioning in volapi much easier later.

File: package.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "sdc-clients",
   "description": "Contains node.js client libraries for SDC REST APIs.",
-  "version": "9.2.0",
+  "version": "9.3.0",
   "homepage": "http://www.joyent.com",
   "repository": {
```
Comment (trentm) [inline package.json]: Should update assert-plus to `1.*` latest, it has fixes for a number of methods, esp. the "optional*" ones.

File: lib/volapi.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        if (options) {
+            if (options.headers) {
+                reqOpts.headers = options.headers;
+            }
+
+            reqOpts.log = options.log || this.log;
+        }
+
+        if (params.context) {
+            reqOpts.headers['x-context'] = JSON.stringify(params.context);
```
Comment (trentm) [inline lib/volapi.js]: Do we know that this is needed and gets carried through? Assuming this is for either MachineAudit in cloudapi and/or for RBAC?

File: test/volapi.test.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2016, Joyent, Inc.
+ */
+
+var VOLAPI = require('../lib/index').VOLAPI;
```
Comment (trentm) [inline test/volapi.test.js]: Local lib imports in a separate block below the external package imports.

File: test/volapi.test.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+var restify = require('restify');
+
+var VOLAPI_IP = process.env.VOLAPI_IP || '10.99.99.70';
+var VOLAPI_URL = 'http://' + VOLAPI_IP;
+
+var volApiClient;
+
+exports.setUp = function (callback) {
+    var logger = new bunyan.createLogger({
+            name: 'vapi_unit_test',
```
Comment (trentm) [inline test/volapi.test.js]: s/vapi/volapi/

Call the var `log` instead of `logger` perhaps? Suspect you just copied other test files here tho.

**Comment by trentm** [comment]
Done my first review. :)

File: lib/volapi.js (hunk 1 above)
Comment (misterdjules) [inline lib/volapi.js]: > Also discussable is whether all client methods should be strict in validating input options. I'm leaning towards yes.

Do you mean strictly validating the properties of the `options` object passed to `createVolume` and all other exposed methods?

File: lib/volapi.js (hunk 2 above)
Comment (misterdjules) [inline lib/volapi.js]: Agreed, will fix.

File: lib/volapi.js (hunk 3 above)
Comment (misterdjules) [inline lib/volapi.js]: Sounds good, will fix.

File: lib/volapi.js (hunk 4 above)
Comment (misterdjules) [inline lib/volapi.js]: Sounds good, will fix.

File: package.json (hunk 5 above)
Comment (misterdjules) [inline package.json]: Good point, will do.

File: lib/volapi.js (hunk 6 above)
Comment (misterdjules) [inline lib/volapi.js]: I assumed this was used by RBAC, and saw it was setup in the same way in the APIs for other services, but I'll double check.

File: test/volapi.test.js (hunk 7 above)
Comment (misterdjules) [inline test/volapi.test.js]: Will fix.

File: test/volapi.test.js (hunk 8 above)
Comment (misterdjules) [inline test/volapi.test.js]: Will fix.

**Comment by misterdjules** [comment]
@trentm Addressed most of your comments except for:

* https://github.com/joyent/node-sdc-clients/pull/8#discussion_r59661691: we still need to have that discussion.
* https://github.com/joyent/node-sdc-clients/pull/8#discussion_r59662138: it's actually not clear why the API would need that now, so I removed it for now, we can also discuss that and re-add it later if that's relevant.

File: lib/volapi.js (hunk 1 above)
Comment (trentm) [inline lib/volapi.js]: Yes, that's what I meant. E.g. https://github.com/joyent/sdcadm/blob/f7e0a0eeedb8666032a3151caa9a2a31736ae3fe/lib/common.js#L40-L75

**Comment by misterdjules** [comment]
Work on this has been happening in the `tritonnfs` feature branch, and will be reviewed with Gerrit at cr.joyent.us.
